packages/loopring_v3/contracts/amm/AmplifiedAmmController.sol

```diff
@@ -0,0 +1,140 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright 2017 Loopring Technology Limited.
+pragma solidity ^0.7.0;
+pragma experimental ABIEncoderV2;
+
+import "../amm/LoopringAmmPool.sol";
+import "../lib/Claimable.sol";
+import "../lib/MathUint.sol";
+import "../lib/MathUint96.sol";
+import "../thirdparty/SafeCast.sol";
+import "./IAmmController.sol";
+
+
+/// @author Brecht Devos - <brecht@loopring.org>
+contract AmplifiedAmmController is IAmmController, Claimable
+{
+    using MathUint          for uint;
+    using MathUint96        for uint96;
+    using SafeCast          for uint;
+
+    uint public constant AMPLIFICATION_FACTOR_BASE = (10 ** 18);
+
+    uint public constant CURVE_CHANGE_MIN_DELAY   = 7 days;
+    uint public constant CURVE_CHANGE_AUTH_WINDOW = 7 days;
+
+    mapping (address => uint) public amplificationFactors;
+    mapping (address => uint) public curveChangeAuthorization;
+
+    function getInitialVirtualBalances(
+        uint96[] memory joinAmounts
+        )
+        external
+        view
+        override
+        returns (uint96[] memory)
+    {
+        uint amplificationFactor = getAmplificationFactor(msg.sender);
+        uint96[] memory vTokenBalancesL2 = new uint96[](joinAmounts.length);
+        for (uint i = 0; i < joinAmounts.length; i++) {
+            vTokenBalancesL2[i] = (uint(joinAmounts[i]).mul(amplificationFactor) / AMPLIFICATION_FACTOR_BASE).toUint96();
+            require(vTokenBalancesL2[i] > 0, "ZERO_VIRTUAL_BALANCE");
+        }
+        return vTokenBalancesL2;
+    }
+
+   function authorizeVirtualBalances(
+        uint96[] memory balances,
+        uint96[] memory /*vBalancesOld*/,
+        uint96[] memory vBalancesNew,
+        bytes    memory /*data*/
+        )
+        external
+        override
+        returns (bool)
+    {
+        address pool = msg.sender;
+
+        // Check if a curve change was explicitly authorized
+        if (consumeCurveChangeAuthorized(pool)) {
+            return true;
+        }
+
+        if (getAmplificationFactor(pool) != AMPLIFICATION_FACTOR_BASE) {
```

Comment:
How about we change `if (getAmplificationFactor(pool) != AMPLIFICATION_FACTOR_BASE) ` to `if (amplificationFactors[pool] != 0) ` so we only give special logic to uninitialized pools instead of AF=1 pools?

Reply:
I think that works for how we want to use it. But if the AF = 1 then the virtual balances should already match the actual balances so in practice it shouldn't make any difference.